salvo/src/lib/test_source_tree.py

```diff
@@ -9,83 +9,240 @@
      operation.
   """
-  pass
+  origin = 'https://github.com/someawesomeproject/repo.git'
+  st = _generate_source_tree_from_origin(origin)
+
+  mock_run_command.side_effect = mock_run_command_side_effect_failure
+
+  git_status = 'git status'
+  git_clone = 'git clone https://github.com/someawesomeproject/repo.git .'
+  cmd_params = cmd_exec.CommandParameters(cwd=mock.ANY)
+
+  result = st.pull()
+  assert not result
+
+  calls = [
+      mock.call(git_status, cmd_params),
+      mock.call(git_clone, cmd_params)
+  ]
+  mock_run_command.assert_has_calls(calls)
+
+  assert st.get_origin() == origin
 
 def test_retrieve_head_hash():
   """Verify that we can determine the hash for the head commit."""
-  pass
+
+  origin = 'https://github.com/someawesomeproject/repo.git'
+  st = _generate_source_tree_from_origin(origin)
+
+  gitcmd = ("git rev-list --no-merges --committer='GitHub <noreply@github.com>'"
+            " --max-count=1 HEAD")
+  git_output = "some_long_hex_string_that_is_the_hash"
+
+  with mock.patch('src.lib.cmd_exec.run_command',
+                  mock.MagicMock(return_value=git_output)) as magic_mock:
+    hash_string = st.get_head_hash()
+    cmd_params = cmd_exec.CommandParameters(cwd=mock.ANY)
+    magic_mock.assert_called_once_with(gitcmd, cmd_params)
+
+    assert hash_string == git_output
 
 def _check_output_side_effect(*args):
   """Respond to the mocked function call with the correct output.
 
   Args:
     args: the arguments supplied to the mock.  The first argument is the
       command being executed.
-
-    kwargs: the keyword arguments supplied to the mock. In this case
-      the 'cwd' (working directory) is the one keyword that should always
-      be present
   """
-  pass
+
+  kwargs = args[1]._asdict()
+  assert 'cwd' in kwargs
+  assert kwargs['cwd']
+
+  if args[0] == 'git remote -v':
+    return 'origin https://github.com/someawesomeproject/repo.git (fetch)'
+
+  elif args[0] == \
+      "git clone https://github.com/someawesomeproject/repo.git .":
+    return "Cloning into \'.\'"
+
+  elif args[0] == ("git rev-list --no-merges "
+                   "--committer=\'GitHub <noreply@github.com>\' "
+                   "--max-count=2 fake_commit_hash_1"):
+    return ("fake_commit_hash_1\n"
+            "fake_commit_hash_2\n")
+
+  elif args[0] == "git status":
+    return "Your branch is up to date with \'some_random_branch\'"
+
+  raise NotImplementedError("Unhandled argument in side effect: %s" % args)
 
 @mock.patch('src.lib.cmd_exec.run_command')
 def test_get_previous_commit(mock_check_output):
-  """Verify that we can identify one commit prior to a specified hash. """
-  pass
+  """
+    Verify that we can identify one commit prior to a specified hash.
+    """
+  origin = 'https://github.com/someawesomeproject/repo.git'
+  st = _generate_source_tree_from_origin(origin)
+
+  commit_hash = 'fake_commit_hash_1'
+
+  mock_check_output.side_effect = _check_output_side_effect
```

Comment:
A question on complexity - do we need to define `_check_output_side_effect()` as a separate function which combines all possible outputs? This makes the reader scroll up and down from the test case to the function and requires the reader to understand the side_effect function.

Can we just set the return value here in the test so the reader has everything in one location? E.g. just set `mock_check_output.return_value`.

Seems like that would allow us to remove some of the helper code and keep test cases self-contained.

Applies to `_check_output_side_effect_fail` as well.

Reply:
Guess I got a little happy witht he mocks.  I've removed `_check_output_side_effect_fail` and used `return_value` in this case.  I also consoliated the side effect functions as much as I could.